planner/core/rule_build_key_info.go

```diff
@@ -209,35 +209,45 @@ func (p *LogicalJoin) BuildKeyInfo(selfSchema *expression.Schema, childSchema []
 }
 
 // checkIndexCanBeKey checks whether an Index can be a Key in schema.
-func checkIndexCanBeKey(idx *model.IndexInfo, columns []*model.ColumnInfo, schema *expression.Schema) expression.KeyInfo {
+func checkIndexCanBeKey(idx *model.IndexInfo, columns []*model.ColumnInfo, schema *expression.Schema) (uniqueKey, newKey expression.KeyInfo) {
 	if !idx.Unique {
-		return nil
+		return nil, nil
 	}
-	newKey := make([]*expression.Column, 0, len(idx.Columns))
-	ok := true
+	newKeyOK := true
+	uniqueKeyOK := true
 	for _, idxCol := range idx.Columns {
 		// The columns of this index should all occur in column schema.
 		// Since null value could be duplicate in unique key. So we check NotNull flag of every column.
-		find := false
+		findNewKey := false
+		findUniqueKey := false
 		for i, col := range columns {
 			if idxCol.Name.L == col.Name.L {
-				if !mysql.HasNotNullFlag(col.Flag) {
+				uniqueKey = append(uniqueKey, schema.Columns[i])
+				findUniqueKey = true
+				if newKeyOK {
+					if !mysql.HasNotNullFlag(col.Flag) {
+						newKeyOK = false
+						break
+					}
+					newKey = append(newKey, schema.Columns[i])
+					findNewKey = true
 					break
 				}
-				newKey = append(newKey, schema.Columns[i])
-				find = true
-				break
 			}
 		}
-		if !find {
-			ok = false
+		if !findNewKey && !findUniqueKey {
```

Comment:
```suggestion
		if !findUniqueKey {
```

`!findUniqueKey` implies `!findNewKey`.

Reply:
done